joomla/joomla-cms
# [4.0] Deliver ES2015+ javascript

Pull Request for Issue # .

### Summary of Changes

Use `type=module` for ES2017 and `nomodule` for ES5 scripts

### Testing Instructions

Test every url Front End  and Back End to confirm that there's no console errors and all the interactive elements continue to work as before
Especially check the Admin login by insering a username but ignoring the password should raise a message (formbvalidation)
Set the editor to Codemirror and create an article and also check that editing a template file is working as before

### Actual result BEFORE applying this Pull Request



### Expected result AFTER applying this Pull Request



### Documentation Changes Required

Just as a comment here: all `.es6.js` and `.w-c.es6.js` are handled by Rollup, meaning that any `import ` will be correctly bundled. Also the compiled `-es5.js` is again created by Rollup using Babel. This adds a delay on the build command BUT now all the ES6 files are ES2017 (meaning there is translation to a minimum version) against the previous state (ES6 files 
could have eg 2021 code that would be problematic for many browsers as the coverage is way bellow the 95% threshold, which is assumed the industry standard for the compatibility lower boundary. Worth watching https://www.youtube.com/watch?v=cLxNdLK--yI by @developit and @housseindjirdeh as this PR basically implements exactly that)

### Performance

The impact could be from 1s to 20s depending on the actual hardware

#### Before

The CI server
![Screenshot 2021-02-25 at 13 26 35](https://user-images.githubusercontent.com/3889375/109153630-726fc400-776d-11eb-80b1-cad597d7658f.png)

My mac
![Screenshot 2021-02-25 at 22 22 32](https://user-images.githubusercontent.com/3889375/109219604-18482080-77b8-11eb-9b80-a1d20dd8001d.png)


#### After

The CI server
![Screenshot 2021-02-25 at 13 28 16](https://user-images.githubusercontent.com/3889375/109153668-7c91c280-776d-11eb-9e6a-ee16a62a109a.png)
 
My mac
![Screenshot 2021-02-25 at 22 20 23](https://user-images.githubusercontent.com/3889375/109219642-25650f80-77b8-11eb-8ddf-66dc6b1414b6.png)


Calling for testing @richard67 @Fedik @anibalsanchez @alikon @laoneo 

> how Joomla is doing the transition

- Started with the bootstrap PR: https://github.com/joomla/joomla-cms/pull/31990
- Then transitioned the Web Components/Custom Elements https://github.com/joomla/joomla-cms/pull/32171
- Finally this PR transitions all the remaining scripts

> what an extension developer should do to follow the steps

- If the dev is using Web Assets then the change is a simple switch of the definitions of the scripts (assuming that the source is modern JS and there is a bundler capable of producing ES2017 and ES5 scripts, continue reading).
Eg legacy script is, let's say some jQuery, so:

```json
    {
       "name": "com_example.awesome-script",
       "type": "script",
       "uri": "com_example/awesome-script.min.js",
       "dependencies": [
         "core"
       ],
       "attributes": {
         "defer": true
       }
``` 

Modern way assumes that the script has a source (modern JS) that transpiles to 2 files, one ES2017 and one ES5 (the one with -es5):

```json
    {
       "name": "com_example.awesome-script.es5",
       "type": "script",
       "uri": "com_example/awesome-script-es5.min.js",
       "dependencies": [
         "core"
       ],
       "attributes": {
         "nomodule": true,
        "defer": true
       },
    }
    {
       "name": "com_example.awesome-script",
       "type": "script",
       "uri": "com_example/awesome-script.min.js",
       "dependencies": [
         "com_example.awesome-script.es5"
       ],
       "attributes": {
         "type": "module"
       }
  }
``` 

- If the dev is using the HTMLHelpr then the change is just an additional call.
Eg: Legacy: 

```php
HTMLHelper::_('script', 'com_example/awesome-script.min.js', ['version' => 'auto', 'relative' => true], ['defer' => true]);
```

Eg Modern:

```php
HTMLHelper::_('script', 'com_example/awesome-script-es5.min.js', ['version' => 'auto', 'relative' => true], ['defer' => true, 'nomodule' => true]);
HTMLHelper::_('script', 'com_example/awesome-script.min.js', ['version' => 'auto', 'relative' => true], ['type' => 'module']);
```

Mind that the legacy script `com_example/awesome-script.min.js` was theoretically rewritten as modern JS and was transpiled to ES2017 with the same name and then a transpiled ES5 was produced by a bundler and had a theoretical name `com_example/awesome-script-es5.min.js`. But this is the convention Joomla is using, devs can use their own naming convention here, Joomla will not impose anything in this part.

> What the CMS provides

Joomla has a set of tools for compiling a modern script to ES2017 and ES5 files. Devs can use Rollup, Webpack, Parcel, ESBuild, SnowPack, or any other bundler out there. The scripts for the bundling live at https://github.com/joomla/joomla-cms/tree/4.0-dev/build/build-modules-js/javascript and devs can even use them as-is if they develop on top of the Joomla repo (this is probably an antipattern, but...) or fork them for their own benefit.
